Matched 1058 tokens.
============= Module ===========
== Sub Tree ==
class  ChatTest
  Fields: 
    listeningPort=0
  Instance Initializer: 
  Constructors: 
  Methods: 
    func  main(args)  throws:     Throwable
      testStartStop()
      testPortOpen()
      testAsksForName()
      testUseName()
      testConnectDisconnectConnect()
      testUsernameAndMessage()
      testDontReceiveMessageInNameState()
    func  startServer()  throws:     IOException
      Decl: server=new ChatServer(0)
      Decl: address=(InetSocketAddress)server.getSocketAddress()
      listeningPort Assign address.getPort()
      server.run()
      return server
    func  testStartStop()  throws:     Exception
      Decl: server=startServer()
      server.shutdown()
    func  testPortOpen()  throws:     Exception
      Decl: server=startServer()
      Decl: socket=new Socket("localhost",listeningPort)
      cond-branch cond:socket.isConnected()
      true branch :
        new RuntimeException("Failed to connect to server: port not open")
      false branch :


      server.shutdown()

    func  testAsksForName()  throws:     Exception
      Decl: server=startServer()
      Decl: socket=new Socket("localhost",listeningPort)
      Decl: reader=new BufferedReader(new InputStreamReader(socket.getInputStream()))
      Decl: string=readAvailableString(reader)
      cond-branch cond:string.equals("Name: ")
      true branch :
        new RuntimeException("Server doesn't send Name: ")
      false branch :


      server.shutdown()

    func  testUseName()  throws:     Throwable
      Decl: server=startServer()
      performTestUseName()

      server.shutdown()

    func  testConnectDisconnectConnect()  throws:     Exception
      Decl: server=startServer()
      performTestConnectDisconnectConnect()

      server.shutdown()

    func  testUsernameAndMessage()  throws:     Exception
      Decl: server=startServer()
      performTestUsernameAndMessage()

      server.shutdown()

    func  testDontReceiveMessageInNameState()  throws:     Exception
      Decl: server=startServer()
      performDontReceiveMessageInNameState()

      server.shutdown()

    func  assertEqual(exception,value,expected)  throws: 
      cond-branch cond:expected EQ value
      true branch :
        return 
      false branch :

      cond-branch cond:expected EQ null
      true branch :
        exception.add(new RuntimeException("Expected null, but was: " Add value))
        return 
      false branch :

      cond-branch cond:expected.equals(value)
      true branch :
        exception.add(new RuntimeException("Expected: " Add expected Add " but was: " Add value))
        return 
      false branch :

    func  performDontReceiveMessageInNameState()  throws:     Exception
      Decl: barrier1=new CyclicBarrier(2)
      Decl: barrier2=new CyclicBarrier(2)
      Decl: barrier3=new CyclicBarrier(2)
      Decl: exceptions=Collections.synchronizedList(new ArrayList())
      Decl: chatConnection=new ChatConnection()
      Decl: client2=new Thread(new ChatConnection())
      client2.start()
      chatConnection.run()
      cond-branch cond:exceptions.isEmpty()
      true branch :
        exceptions.get(0)
      false branch :

    func  waitForJoin(reader,s)  throws:     IOException
      Decl: joined
      do         joined Assign readAvailableString(reader)
while (joined NE null Land joined.contains("Welcome " Add s))
    func  performTestUsernameAndMessage()  throws:     Exception
      Decl: barrier1=new CyclicBarrier(2)
      Decl: barrier2=new CyclicBarrier(2)
      Decl: barrier3=new CyclicBarrier(2)
      Decl: exceptions=Collections.synchronizedList(new ArrayList())
      Decl: chatConnection=new ChatConnection()
      Decl: client2=new Thread(new ChatConnection())
      client2.start()
      chatConnection.run()
      cond-branch cond:exceptions.isEmpty()
      true branch :
        exceptions.get(0)
      false branch :

  LocalClasses: 
  LocalInterfaces: 

